cuda/matrix/ell_kernels.cu

```diff
@@ -67,40 +76,60 @@ __device__ void spmv_kernel(const size_type num_rows,
 {
     const auto tidx =
         static_cast<IndexType>(blockDim.x) * blockIdx.x + threadIdx.x;
-
-    if (tidx < num_rows) {
+    const auto nwarps_per_row =
+        gridDim.x * blockDim.x / num_rows / subwarp_size;
+    const auto x = tidx / subwarp_size / nwarps_per_row;
+    const auto warp_id = tidx / subwarp_size % nwarps_per_row;
+    const auto y_start = tidx % subwarp_size +
+                         num_stored_elements_per_row * warp_id / nwarps_per_row;
+    const auto y_end =
+        num_stored_elements_per_row * (warp_id + 1) / nwarps_per_row;
+    if (x < num_rows) {
+        const auto tile_block =
+            group::tiled_partition<subwarp_size>(group::this_thread_block());
         ValueType temp = zero<ValueType>();
         const auto column_id = blockIdx.y;
-        for (IndexType idx = 0; idx < num_stored_elements_per_row; idx++) {
-            const auto ind = tidx + idx * stride;
+        for (IndexType idx = y_start; idx < y_end; idx += subwarp_size) {
+            const auto ind = x + idx * stride;
             const auto col_idx = col[ind];
             if (col_idx < idx) {
                 break;
             } else {
                 temp += val[ind] * b[col_idx * b_stride + column_id];
             }
         }
-        c[tidx * c_stride + column_id] =
-            op(temp, c[tidx * c_stride + column_id]);
+        const auto answer = reduce(
+            tile_block, temp, [](ValueType x, ValueType y) { return x + y; });
+        if (tile_block.thread_rank() == 0) {
+            if (atomic) {
+                atomic_add(&(c[x * c_stride + column_id]),
+                           op(answer, c[x * c_stride + column_id]));
+            } else {
+                c[x * c_stride + column_id] =
+                    op(answer, c[x * c_stride + column_id]);
+            }
+        }
     }
 }
 
 
-template <typename ValueType, typename IndexType>
+template <int subwarp_size, bool atomic = false, typename ValueType,
+          typename IndexType>
 __global__ __launch_bounds__(default_block_size) void abstract_spmv(
```

Comment:
We are not always consistent with this, but sometimes we put all the kernels in an extra namespace `kernel`. See `jacobi_kernels.cu` for example.